pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,392 @@
+				if c.err != nil {
+					return
+				}
+			} else {
+				c.err = c.sendFlushToRemote(ctx)
+				if c.err != nil {
+					return
+				}
+				task.done(0)
+			}
+			// Reset the ticker when a new task is received. Avoid sending empty chunks too frequently.
+			ticker.Reset(updateFlushedChunkDuration)
+		case <-ticker.C:
+			// Periodically send empty chunks to update the flushed chunkID.
+			if c.getFlushedChunkID() == c.getLastChunkID() {
+				continue
+			}
+			c.err = c.putEmptyChunk(ctx)
+			if c.err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (c *chunkSender) putEmptyChunk(ctx context.Context) error {
+	chunk := &chunk{id: c.getLastChunkID(), data: nil}
+	return c.putChunkToRemote(ctx, chunk)
+}
+
+func (c *chunkSender) putChunkToRemote(ctx context.Context, chunk *chunk) error {
+	// cache chunk data
+	if len(chunk.data) != 0 {
+		err := c.chunksCache.put(chunk.id, chunk.data)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	url := fmt.Sprintf(putChunkURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id, chunk.id)
+
+	data, err := sendRequest(ctx, c.httpClient, "PUT", url, chunk.data)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	result := new(PutChunkResult)
+	err = json.Unmarshal(data, result)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = c.handlePutChunkResult(ctx, result, chunk.id)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+func (c *chunkSender) handlePutChunkResult(ctx context.Context, result *PutChunkResult, expectedChunkID uint64) error {
+	if result.Canceled {
+		c.e.logger.Error("failed to put chunk, task is canceled",
+			zap.String("error", result.Error),
+			zap.Bool("finished", result.Finished))
+		return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+	}
+
+	if result.HandledChunkID != expectedChunkID {
+		c.e.logger.Info("remote worker may restart, retry to put chunk",
+			zap.Uint64("expected ChunkID", expectedChunkID),
+			zap.Uint64("handled ChunkID", result.HandledChunkID),
+			zap.Uint64("flushed ChunkID", result.FlushedChunkID))
+
+		nextChunkID := result.HandledChunkID + 1
+		for nextChunkID <= expectedChunkID {
+			url := fmt.Sprintf(putChunkURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id, nextChunkID)
+
+			buf, err := c.chunksCache.get(nextChunkID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			c.e.logger.Info("retry to put chunk",
+				zap.Uint64("sender", c.id),
+				zap.Uint64("chunkID", nextChunkID))
+
+			data, err := sendRequest(ctx, c.httpClient, "PUT", url, buf)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			result = new(PutChunkResult)
+			err = json.Unmarshal(data, result)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			nextChunkID = result.HandledChunkID + 1
+		}
+	}
+
+	state := c.state.Load()
+	lastFlushedChunkID := state.FlushedChunkID + 1
+	for lastFlushedChunkID <= result.FlushedChunkID {
+		err := c.chunksCache.clean(lastFlushedChunkID)
```

Comment:
WIll the remote worker restarts again after we clean the chunk? Please add comments to explain the communication protocol and guarantees